### PR TITLE
fix: Optimize some functions to rewrite dictionary-encoded strings

### LINKED ISSUE #504: perf: Optimize some functions to rewrite dictionary-encoded strings
### What is the problem the feature request solves?

We have some functions that unpack dictionary-encoded strings and then perform a transformation on the unpacked array, but it would be better to transform the dictionary instead. We should optimize these functions to do that.

The functions include:

- hex
- cast string to other types

There may be others.

### Describe the potential solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by vaibhawvipul**
## Which issue does this PR close?

Closes #504 .

## Rationale for this change

Improves performance

## What changes are included in this PR?

rewrite dictionary encoded strings

## How are these changes tested?

All test cases pass.

### CODE REVIEW

**Comment by andygrove** [comment]
Thanks @vaibhawvipul. This is looking good. There is also this section in `cast.rs`:

```rust
    fn cast_array(&self, array: ArrayRef) -> DataFusionResult<ArrayRef> {
        let to_type = &self.data_type;
        let array = array_with_timezone(array, self.timezone.clone(), Some(to_type))?;
        let from_type = array.data_type().clone();

        // unpack dictionary string arrays first
        // TODO: we are unpacking a dictionary-encoded array and then performing
        // the cast. We could potentially improve performance here by casting the
        // dictionary values directly without unpacking the array first, although this
        // would add more complexity to the code
```

**Comment by vaibhawvipul** [comment]
> Thanks @vaibhawvipul. This is looking good. There is also this section in `cast.rs`:
> 
> ```rust
>     fn cast_array(&self, array: ArrayRef) -> DataFusionResult<ArrayRef> {
>         let to_type = &self.data_type;
>         let array = array_with_timezone(array, self.timezone.clone(), Some(to_type))?;
>         let from_type = array.data_type().clone();
> 
>         // unpack dictionary string arrays first
>         // TODO: we are unpacking a dictionary-encoded array and then performing
>         // the cast. We could potentially improve performance here by casting the
>         // dictionary values directly without unpacking the array first, although this
>         // would add more complexity to the code
> ```

This is done @andygrove . Thank you for pointing this out.

**Comment by andygrove** [comment]
There is a test failure:

```
Cause: org.apache.comet.CometNativeException: Unsupported data type Dictionary(Int32, Binary) for function md5
```

**Comment by vaibhawvipul** [comment]
> There is a test failure:
> 
> ```
> Cause: org.apache.comet.CometNativeException: Unsupported data type Dictionary(Int32, Binary) for function md5
> ```

fixed

File: native/core/src/execution/datafusion/expressions/cast.rs
Comment (andygrove) [inline native/core/src/execution/datafusion/expressions/cast.rs:522]: This code looks good and is converting the input dictionary array to the output dictionary array that we want

File: native/core/src/execution/datafusion/expressions/cast.rs
Comment (andygrove) [inline native/core/src/execution/datafusion/expressions/cast.rs:114]: It looks like there is no functional change here and this seems unrelated to the goal of the PR. I think it would be better to remove this change from this PR.

File: native/core/src/execution/datafusion/expressions/cast.rs
Comment (vaibhawvipul) [reply native/core/src/execution/datafusion/expressions/cast.rs:521]: hmm.. i have one question. Is it correct to expect dictionary as an output? Shouldn't the output be equal to what to_type is set? `DataType::Timestamp(TimeUnit::Microsecond, Some(timezone.clone().into()))` i.e TimestampArray?

Comment (andygrove) [reply native/core/src/execution/datafusion/expressions/cast.rs:521]: Ah, yes, that is a good point. I may have misundestood the goal here. Let me take another look.

Comment (andygrove) [reply native/core/src/execution/datafusion/expressions/cast.rs:521]: I can see that my test as written makes no sense now. It would be good to add a modified version of this test though for the current behavior.

File: native/core/src/execution/datafusion/expressions/scalar_funcs/hex.rs
Comment (vaibhawvipul) [inline native/core/src/execution/datafusion/expressions/scalar_funcs/hex.rs:149]: did a small benchmark by timing `hex` testcase in CometExpressionSuite. 

Old code - 3600 ms
New code - 2950ms

File: native/core/src/execution/datafusion/expressions/cast.rs
Comment (andygrove) [inline native/core/src/execution/datafusion/expressions/cast.rs:520]: I don't know if we would currently hit the use case of casting *to* a dictionary type, or whether this is something that could help us in the future (it seems like we could take advantage of this when running natively even though Spark doesn't support keeping data dictionary-encoded after a cast, as far as I know, but maybe @viirya could confirm).

File: native/core/src/execution/datafusion/expressions/cast.rs
Comment (andygrove) [inline native/core/src/execution/datafusion/expressions/cast.rs:1726]: Could you also add a version of this test that casts to a regular timestamp (not dictionary encoded), since that is the primary use case.

**Review by andygrove — APPROVED** [review]
LGTM. Thanks for refactoring the code @vaibhawvipul, it is looking much cleaner now. I would like to see the additional test that I mentioned before we merge this.

File: native/core/src/execution/datafusion/expressions/cast.rs
Comment (vaibhawvipul) [reply native/core/src/execution/datafusion/expressions/cast.rs]: Hey this is named incorrectly. I have changed it.

This test actually tests 'dict to timstamp array' cast.

File: native/core/src/execution/datafusion/expressions/cast.rs
Comment (vaibhawvipul) [reply native/core/src/execution/datafusion/expressions/cast.rs:520]: awaiting @viirya 's comment before making any change.

**Comment by andygrove** [comment]
I think this is ready to merge, but I'd like to get a another review. Perhaps @parthchandra @huaxingao @kazuyukitanimura or @viirya could take a look

File: native/spark-expr/src/cast.rs
Comment (parthchandra) [inline native/spark-expr/src/cast.rs:735]: I don't see a logic change here. Is this just refactoring?

Comment (vaibhawvipul) [reply native/spark-expr/src/cast.rs:735]: yes, no logic change here.

**Review by parthchandra — APPROVED** [review]
lgtm

**Comment by andygrove** [comment]
Thanks @vaibhawvipul and thanks for the review @parthchandra

**Pull request merged**
